timelock-corruption-detection/src/test/java/com/palantir/history/SqliteHistoryQueryTest.java

```diff
@@ -0,0 +1,123 @@
+import com.palantir.paxos.SqlitePaxosStateLog;
+
+public class SqliteHistoryQueryTest {
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder();
+
+    private static final Client CLIENT = Client.of("tom");
+    private static final String USE_CASE_LEARNER = "useCase!learner";
+    private static final String USE_CASE_ACCEPTOR = "useCase!acceptor";
+
+    private DataSource dataSource;
+    private Jdbi jdbi;
+    private PaxosStateLog<PaxosValue> learnerLog;
+    private PaxosStateLog<PaxosAcceptorState> acceptorLog;
+    private LocalHistoryLoader history;
+
+    @Before
+    public void setup() {
+        dataSource = SqliteConnections.getPooledDataSource(tempFolder.getRoot().toPath());
+        learnerLog = SqlitePaxosStateLog.create(
+                ImmutableNamespaceAndUseCase.of(CLIENT, USE_CASE_LEARNER), dataSource);
+        acceptorLog = SqlitePaxosStateLog.create(
+                ImmutableNamespaceAndUseCase.of(CLIENT, USE_CASE_ACCEPTOR), dataSource);
+
+        history = LocalHistoryLoader.create(SqlitePaxosStateLogHistory.create(dataSource));
+    }
+
+    @Test
+    public void canGetAllLearnerLogsSince() {
+        IntStream.range(0, 100).forEach(i -> writeValueForLogAndRound(learnerLog, i + 1));
+        LearnerAndAcceptorRecords learnerAndAcceptorRecords
+                = history.loadLocalHistory(ImmutableNamespaceAndUseCase.of(
+                        CLIENT, UseCaseUtils.getPaxosUseCasePrefix(USE_CASE_LEARNER)), 5L);
+        assertThat(learnerAndAcceptorRecords.acceptorRecords().size()).isEqualTo(0);
+        assertThat(learnerAndAcceptorRecords.learnerRecords().size()).isEqualTo(95);
+    }
+
+    @Test
+    public void canGetAllAcceptorLogsSince() {
+        IntStream.range(0, 100).forEach(i -> writeAcceptorStateForLogAndRound(acceptorLog, i + 1));
+        LearnerAndAcceptorRecords learnerAndAcceptorRecords
+                = history.loadLocalHistory(ImmutableNamespaceAndUseCase.of(
+                CLIENT, UseCaseUtils.getPaxosUseCasePrefix(USE_CASE_LEARNER)), 5L);
+        assertThat(learnerAndAcceptorRecords.learnerRecords().size()).isEqualTo(0);
+        assertThat(learnerAndAcceptorRecords.acceptorRecords().size()).isEqualTo(95);
+    }
+
+    @Test
+    public void canGetAllUniquePairsOfNamespaceAndClient() {
+        IntStream.range(0, 100).forEach(i -> {
+            PaxosStateLog<PaxosValue> otherLog
+                    = SqlitePaxosStateLog.create(ImmutableNamespaceAndUseCase.of(Client.of("client" + i),
+                    USE_CASE_LEARNER), dataSource);
+            writeValueForLogAndRound(otherLog, 1L);
+        });
+        Set<NamespaceAndUseCase> allNamespaceAndUseCaseTuples = SqlitePaxosStateLogHistory.create(
+                dataSource).getAllNamespaceAndUseCaseTuples();
+        assertThat(allNamespaceAndUseCaseTuples.size()).isEqualTo(100);
+    }
+
+    private PaxosValue writeValueForLogAndRound(PaxosStateLog<PaxosValue> log, long round) {
```

Comment:
These tests are pretty decent, maybe add a test where some rounds are not present and you have records in both. Maybe even verify the actual paxosvalue

Reply:
done